lib/analyzer/image-inspector.ts

```diff
@@ -12,6 +18,83 @@ async function detect(
   return JSON.parse(info.stdout)[0];
 }
 
+/**
+ * TODO: This function could probably have a better name and location
+ *
+ * In the case that an `ImageType.Identifier` is detected we need to produce
+ * an image archive, either by saving the image if it's already loaded into
+ * the local docker daemon, or by pulling the image from a remote registry and
+ * saving it to the filesystem directly.
+ *
+ * Users may also provide us with a URL to an image in a Docker compatible
+ * remote registry.
+ *
+ * @param {string} targetImage
+ * @param {string} [username]
+ * @param {string} [password]
+ */
+async function getArchiveLocation(
+  targetImage: string,
+  username?: string,
+  password?: string,
+): Promise<string> {
+  const docker = new Docker(targetImage);
+
+  try {
+    // TODO: Eventually need to provide support for specifying a user-defined
+    // temporary directory. Defaulting to `path.join(os.tmpdir(), "image.tar")`
+    // for now as this matches the output from the `snyk-docker-pull` lib.
+    const saveLocation: string = path.join(os.tmpdir(), "image.tar");
+    await docker.save(targetImage, saveLocation);
+    return saveLocation;
+  } catch {
+    debug(
+      `${targetImage} does not exist locally, proceeding to pull image from remote registry.`,
+    );
+  }
+
+  let remainder: string;
+  let hostname: string;
+  let auth: string | undefined;
+
+  // We need to detect if the `targetImage` is part of a URL.
+  // Based on the Docker specification, the hostname should contain
+  // a `.` or `:` before the first instance of a `/` otherwise the
+  // default hostname will be used (registry-1.docker.io).
+  // ref: https://stackoverflow.com/a/37867949
+  const i = targetImage.indexOf("/");
+  if (
+    i === -1 ||
+    (!targetImage.substring(0, i).includes(".") &&
+      !targetImage.substring(0, i).includes(":") &&
+      targetImage.substring(0, i) !== "localhost")
+  ) {
+    hostname = "registry-1.docker.io";
+    remainder = targetImage;
+    const authRequest = await needle(
+      "get",
+      `https://auth.docker.io/token?service=${hostname}&scope=repository:${
+        remainder.split(":")[0]
+      }:pull`,
+      { json: true },
+    );
+    auth = authRequest.body.token;
```

Comment:
Would this be better suited for the `pull-lib`? It feels a bit odd to be outside of the library that deals with auth and registries

Reply:
Yes, ideally this would be handled in the `docker-registry-v2-client`. I have a PR open here: https://github.com/snyk/docker-registry-v2-client/pull/50 which should hopefully make this code redundant.